core/chapters/c12_dictionaries.py

```diff
@@ -0,0 +1,510 @@
+Sorry, you can't do that either. You can only look up a key to get its value, not the other way around.
+
+The dictionary `french` only has 2 keys: `'apple'` and `'box'`. `'pomme'` is a value, not a key.
+
+We'll soon learn why you can't just look up values directly, and what you can do about it.
+
+
+Note that both `french[0]` and `french['pomme']` raised the same type of error: a `KeyError`.
+
+This error means that the provided key (`0` or `'pomme'` in this case) wasn't found in the dictionary.
+
+It's not that `french[0]` isn't *allowed*, it's just that it means the same thing as always:
+find the value associated with the key `0`. In this case it finds that no such key exists.
+
+But `0` *could* be a key, because many types of keys are allowed, including strings and numbers.
+"""
+
+
+class UsingDictionaries(Page):
+    class shopping_cart1(VerbatimStep):
+        """
+Let's see dictionaries in a real life problem. Imagine you're building an online shopping website.
+
+You keep the prices of all your items in a dictionary:
+
+__program_indented__
+        """
+
+        auto_translate_program = False
+
+        program = "prices = {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100}"
+
+
+    class shopping_cart2(VerbatimStep):
+        """
+Here you can see one reason why looking up values in a dictionary could be a problem.
+
+What would `prices[100]` be? `'dog'`? `'cat'`? `['dog', 'cat']`?
+
+The same value can be repeated any number of times in a dictionary.
+
+On the other hand, keys have to be unique. Imagine if your prices started like this:
+
+__program_indented__
+        """
+
+        auto_translate_program = False
+
+        program = "prices = {'apple': 2, 'apple': 3}"
+
+    class shopping_cart3(ExerciseStep):
+        """
+How much does an apple cost? We know it's `prices['apple']`, but is that `2` or `3`?
+
+Clearly there should only be one price, so duplicate keys aren't allowed.
+
+
+Anyway, this is a normal shop where things have one price.
+
+This normal shop has normal customers with normal shopping lists like `['apple', 'box', 'cat']`.
+
+And even though your customers have calculators in their pockets, they still expect you to add up all the prices
+yourself and tell them how much this will all cost, because that's what normal shops do.
+
+So let's write a function that does that. Complete the function below, particularly the line `price = ...`
+
+
+    __copyable__
+    __program_indented__
+
+    def program(self):
+        def total_cost(cart, prices):
+            result = 0
+            for item in cart:
+                price = ... 
+                result += price
+            return result
+
+        assert_equal(
+            total_cost(
+                ['apple', 'box', 'cat'],
+                {'apple': 2, 'box': 5, 'cat': 100, 'dog': 100},
+                ),
+            107,
+        )
+        """ 
+
+        def solution(self):
+            def total_cost(cart: Dict[str, str], prices: Dict[str, str]):
```

Comment:
```suggestion
            def total_cost(cart: List[str], prices: Dict[str, int]):
```

But in any case it doesn't work here to just rely on the type annotations. You need to write a custom `generate_inputs` so that the elements of `cart` are keys in `prices`.